src/ATen/native/xpu/sycl/ReduceSumProdKernels.cpp

```diff
@@ -36,22 +67,56 @@ struct sum_functor {
 };
 
 void sum_kernel(TensorIterator& iter) {
-  if (iter.dtype() == kHalf) {
-    return sum_functor<at::Half, float>{}(iter);
-  } else if (iter.dtype(1) == kHalf && iter.dtype() == kFloat) {
-    // type promotion that does cast and reduction in a single kernel
-    return sum_functor<at::Half, float, float>{}(iter);
-  } else if (iter.dtype() == kBFloat16) {
-    return sum_functor<at::BFloat16, float>{}(iter);
-  } else if (iter.dtype(1) == kBFloat16 && iter.dtype() == kFloat) {
-    // type promotion that does cast and reduction in a single kernel
-    return sum_functor<at::BFloat16, float, float>{}(iter);
+  auto general_dispatcher = [](TensorIterator& iter) {
+    AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND2(
+        kBool, kComplexHalf, iter.dtype(), "sum_xpu", [&]() {
+          sum_functor<scalar_t>{}(iter);
+        });
+  };
+  reduce_dispatch<sum_functor>(iter, general_dispatcher);
+}
+
+template <typename acc_t>
+struct NanSumFunctor {
+  inline acc_t operator()(acc_t a, acc_t b) const {
+    return a + (at::_isnan(b) ? acc_t{0.} : acc_t{b});
+  }
+};
+
+template <>
+struct NanSumFunctor<c10::complex<at::Half>> {
+  using scalar_t = c10::complex<at::Half>;
+  using acc_t = at::opmath_type<scalar_t>;
+  inline acc_t operator()(acc_t a, acc_t b) const {
+    return a + (at::_isnan(b) ? acc_t{0.} : acc_t{b});
   }
+};
```

Comment:
Why we don't do as CUDA does?
<img width="569" alt="image" src="https://github.com/user-attachments/assets/98219ee6-2d1a-4fdb-a851-c4a208db7b0b">

Reply:
Why we have specialization for ComplexHalf?